keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,1087 @@
+all storage classes defined for the driver and call `GetCapacity` once
+per class with the parameters defined in the class. The result will be
+stored in one `CSIStoragePoolByClass` per storage class.
+
+If the current combination of topology segment for a pool and storage
+class parameters do not make sense, then a driver must return "zero
+capacity" or an error, in which case external-provisioner will skip
+this combination. This covers the case where some storage class
+parameter selects a certain storage pool, because information will
+then only be recorded for that pool.
+
+When `--enable-capacity=ephemeral` is used, it will call `GetCapacity`
+without parameters and create the special `<ephemeral>`
+`CSIStoragePoolByClass` entry.
+
+While technically these options are orthogonal, not all combinations
+make sense. The expected usage is:
+* A driver that only supports ephemeral volumes should use only
+  `--enable-capacity=ephemeral`. Storage classes that might
+  (accidentally?) be created for the driver will be ignored.
+* A driver where parameters have an effect should use `--enable-capacity=storageclasses`,
+  combined with `--enable-capacity=ephemeral` if ephemeral volumes
+  are supported.
+
+#### CSIStoragePool lifecycle
+
+external-provisioner needs permission to create, update and delete
+`CSIStoragePool` objects. Before creating a new object, it must check
+whether one already exists with the relevant attributes (driver name +
+nodes) and then update that one instead. Obsolete objects needs to be
+removed.
+
+To ensure that `CSIStoragePool` objects get removed when the driver
+deployment gets removed before it has a chance to clean up, each
+`CSIStoragePool` object needs an [owner
+reference](https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#OwnerReference).
+
+For central provisioning, that has to be the deployment or stateful
+set that defines the provisioner pods. That way, provisioning can
+continue seamlessly when there are multiple instances with leadership
+election.
+
+For deployment with a daemon set, making the individual pod the owner
+is better than the daemon set as owner because then other instances do
+not need to figure out when to remove `CSIStoragePool` objects created
+on another node. It is also better than the node as owner because a
+driver might no longer be needed on a node although the node continues
+to exist (for example, via labels).
+
+While external-provisioner runs, it needs to update and potentially
+delete `CSIStoragePool` objects:
+- when nodes change (for central provisioning)
+- when storage classes change (for persistent volumes)
+- when volumes were created or deleted (for central provisioning)
+- when volumes are resized or snapshots are created or deleted (for persistent volumes)
+- periodically, to detect changes in the underlying backing store (all cases)
+
+Because sidecars are currently separated, external-snapshotter is
```

Comment:
s/external-snapshotter/external-provisioner?

Reply:
Yes, thanks for pointing it out.